geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java

```diff
@@ -199,9 +231,12 @@ public void cleanup(boolean fromCacheClientProxy) {
       }
     }
 
-    // Logout the subjects
-    for (final Long subjectId : uniqueIdVsSubject.keySet()) {
-      removeSubject(subjectId);
+    // for integrated security, doesn't matter if this is called from proxy
+    // or from the connection, we are closing the client connection
+    synchronized (this) {
```

Comment:
As Jianxia pointed out earlier, use of synchronization can be avoided by using Concurrent Data structures (CopyOnWriteArrayList); any reason why we can not use them. 
Also, if using synchronized, can the synchronization be done with/on the "uniqueIdVsSubjects" instead on "this".

Reply:
We want to prevent multiple threads calling this block of code and "removeSubject" at the same time, using concurrent data structure wouldn't guarantee that.

if we synchronize on "uniqueIdVsSubjects" here, we can't simply add "synchronize" on the other methods, since the other methods are synchronizing on the "this" object if we put "synchronize" on the method name.